src/faebryk/core/graph.py

```diff
@@ -2,142 +2,47 @@
-        # thus we dont have to merge the graph, just need to setup the ref
-        if LAZY:
-            if not lhs.is_init or not rhs.is_init:
-                if not lhs.is_init:
-                    lhs, rhs = rhs, lhs
-                rhs.links = lhs.links
-                rhs.object = lhs.object
-                lhs.links.add(rhs)
-                rhs._init = True
-                return lhs, True
-
-        if lhs() == rhs():
-            return self, False
-
-        unioned = self._union(self(), other())
-        if unioned is rhs():
-            lhs, rhs = rhs, lhs
-        if unioned is not lhs():
-            lhs.set(unioned)
-
-        res = lhs.link(rhs)
-        if not res:
-            return self, False
-
-        # TODO remove, should not be needed
-        assert isinstance(res.representative, type(self))
+    from faebryk.core.trait import Trait
 
-        return res.representative, True
-
-    def __repr__(self) -> str:
-        G = self()
-        node_cnt = self.node_cnt
-        edge_cnt = self.edge_cnt
-        g_repr = f"{type(G).__name__}({node_cnt=},{edge_cnt=})({hex(id(G))})"
-        return f"{type(self).__name__}({g_repr})"
-
-    @property
-    @abstractmethod
-    def node_cnt(self) -> int: ...
-
-    @property
-    @abstractmethod
-    def edge_cnt(self) -> int: ...
-
-    @abstractmethod
-    def v(self, obj: T): ...
-
-    @abstractmethod
-    def add_edge(self, from_obj: T, to_obj: T, link: "Link"): ...
-
-    # TODO implement everywhere
-    @abstractmethod
-    def remove_edge(self, from_obj: T, to_obj: T | None = None): ...
-
-    @abstractmethod
-    def is_connected(self, from_obj: T, to_obj: T) -> "Link | None": ...
-
-    @abstractmethod
-    def get_edges(self, obj: T) -> Mapping[T, "Link"]: ...
-
-    @staticmethod
-    @abstractmethod
-    def _union(rep: GT, old: GT) -> GT: ...
-
-    def bfs_visit(
-        self,
-        filter: Callable[[list[T], "Link"], bool],
-        start: Iterable[T],
-        G: GT | None = None,
-    ):
-        G = G or self()
-
-        return bfs_visit(
-            lambda n: [
-                o for o, link in self.get_edges(n[-1]).items() if filter(n + [o], link)
-            ],
-            start,
-        )
-
-    def __str__(self) -> str:
-        return f"{type(self).__name__}(V={self.node_cnt}, E={self.edge_cnt})"
-
-    @abstractmethod
-    def __iter__(self) -> Iterator[T]: ...
-
-    # TODO subgraph should return a new GraphView
-    @abstractmethod
-    def subgraph(self, node_filter: Callable[[T], bool]) -> Iterable[T]: ...
+logger = logging.getLogger(__name__)
 
-    def subgraph_type(self, *types: type[T]):
-        return self.subgraph(lambda n: isinstance(n, types))
 
-    # TODO remove boilerplate for lazy, if it becomes a bit more usable
-    def __init_subclass__(cls) -> None:
-        super().__init_subclass__()
-        if LAZY:
-            lazy_construct(cls)
+# TODO move these to C++
+# a bit weird typecheck
+class GraphFunctions:
```

Comment:
This pattern of a seperate not-quite-subclass to wrap functions is a little odd. Is this only as a placeholder to soon refactor, or do you plan to keep this?

Reply:
Just placeholder to make it quick and easy to refactor old code